File: .NET/Microsoft.Recognizers.Text/Matcher/TrieTree.cs
using System.Collections.Generic;
using System.Linq;

namespace Microsoft.Recognizers.Text.Matcher
{
    public class TrieTree<T> : AbstractMatcher<T>
    {
        public TrieTree()
        {
        }

        protected Node<T> Root { get; private set; } = new Node<T>();

        public override void Insert(IEnumerable<T> value, string id)
        {
            var node = Root;

            foreach (var item in value)
            {
                var child = node[item];

                if (child == null)
                {
                    child = node[item] = new Node<T>();
                }

                node = child;
            }

            node.AddValue(id);
        }

        public override void Init(IEnumerable<T>[] values, string[] ids)
        {
            BatchInsert(values, ids);
            ConvertDictToList(Root);
        }

        public override IEnumerable<MatchResult<T>> Find(IEnumerable<T> queryText)
        {
            var queryArray = queryText.ToArray();
            for (var i = 0; i < queryArray.Length; i++)
            {
                var node = Root;
                for (var j = i; j <= queryArray.Length; j++)
                {
                    if (node.End)
                    {
                        yield return new MatchResult<T>(i, j - i, node.Values);
                    }

                    if (j == queryArray.Length)
                    {
                        break;
                    }

                    var text = queryArray[j];
                    if (node[text] == null)
                    {
                        break;
                    }

                    node = node[text];
                }
            }
        }
    }
}

File: Python/libraries/recognizers-text/recognizers_text/Matcher/trie_tree.py
from typing import List

from recognizers_text import ModelResult

from .node import Node
from .abstract_matcher import AbstractMatcher
from .match_result import MatchResult


class TrieTree(AbstractMatcher):

    def __init__(self):
        self.__root = Node()

    @property
    def root(self) -> Node:
        return self.__root

    def insert(self, value: [], id: str) -> None:
        node = self.root
        for item in value:
            child = node[item]

            if child is None:
                node[item] = Node()
                child = node[item]

            node = child

        node.add_value(id)

    def init(self, values: [], ids: []) -> None:
        self.batch_insert(values, ids)
        list(self.root)

    def find(self, query_text: []) -> []:
        for i in range(0, len(query_text)):
            node = self.root
            j = i
            for j in range(j, len(query_text)+1):
                if node.end:
                    yield MatchResult(i, j - i, node.values)

                if j == len(query_text):
                    break

                text = query_text[j]
                if node[text] is None:
                    break

                node = node[text]
